{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### GFP_AEQV processing"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "import evoVAE.utils.seq_tools as st\n",
    "from evoVAE.utils.webservice import get_sequence_batch\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "pd.set_option(\"display.max_rows\", None)\n",
    "from ete3 import Tree\n",
    "from sequence import *"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Reading in alignment "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Reading the alignment: /Users/sebs_mac/OneDrive - The University of Queensland/honours/data/gfp_alns/independent_runs/no_synthetic/alns/GFP_AEQVI_full_04-29-2022_b08_extants_no_syn.aln\n",
      "Checking for bad characters: ['B', 'J', 'X', 'Z']\n",
      "Performing one hot encoding\n",
      "Number of seqs: 360\n"
     ]
    }
   ],
   "source": [
    "DATA_DIR = '/Users/sebs_mac/OneDrive - The University of Queensland/honours/data/gfp_alns/independent_runs/no_synthetic/alns/'\n",
    "\n",
    "aln = st.read_aln_file(DATA_DIR + 'GFP_AEQVI_full_04-29-2022_b08_extants_no_syn.aln')\n",
    "#st.write_fasta_file(f\"../data/alignments/gfp_alns/GFP_AEQVI_full_04-29-2022_b08.aln\", aln)\n",
    "\n",
    "#draws = []\n",
    "#for draw in range(100):\n",
    "#    draws.append(aln.sample(n=250, random_state=draw, replace=False))\n",
    "\n",
    "#for i, seqs in enumerate(draws):\n",
    "#    st.write_fasta_file(f\"../data/alignments/gfp_alns/GFP_AEQVI_full_04-29-2022_b08_rand_subset_{i}.aln\", seqs)\n",
    "#aln"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Collecting annotations about sequences"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['AEQVI', 'UPI0011C34247', 'UPI0011C34247', 'UPI0011C3426C', 'UPI0011C3426C', 'UPI001C2E920B', 'UPI0000EA3547', 'UPI000B5ABD33', 'UPI000B5ABDBE', 'UPI000B5ABD69', 'UPI000B5ABD53', 'UPI0009795733', 'UPI00135F1855', 'P42212', 'W6KDG8', 'UPI000B5ABDBC', 'UPI0001986815', 'A0A2V2QXC3', 'UPI0002A54BBC', 'Q8GHE4', 'UPI000B5ABD1A', 'UPI000B5ABD2E', 'UPI000067C63B', 'A0A8E6TY54', 'UPI000B5ABD28', 'A0A6S7E9W3', 'UPI000B5DBA18', 'UPI000B5ABD5D', 'A0A5P9VSM6', 'A0A0U3A3D8', 'UPI001C63CC57', 'UPI0001D1F7C5', 'UPI001D0CF09B', 'UPI000C173AFD', 'A0A1V0D974', 'A0A7G8ZY66', 'UPI000B5ABD97', 'UPI00157E3332', 'UPI000B5ABD72', 'UPI001CB8BAD1', 'UPI000B5ABDB8', 'UPI000B8BB2ED', 'UPI0001AB2972', 'UPI00001118CE', 'UPI000B5ABD8B', 'A0A6N6XNT4', 'A0A193CK14', 'UPI000B5ABD25', 'UPI000B5ABD67', 'UPI0018D5A8B2', 'UPI0006891DAB', 'J8VIQ3', 'UPI0010706A2B', 'UPI000F7351CB', 'UPI00157A33EC', 'UPI00027B2CE5', 'UPI0001AB2973', 'UPI000DF72107', 'A0A6M3ZKY0', 'UPI0001986814', 'A0A193DTF7', 'A0A059PIQ0', 'UPI0010A1F85E', 'Q8GHE3', 'UPI0001AF80B1', 'UPI0001BEF289', 'UPI0007414E7C', 'UPI0010FFADAB', 'A0A059PIR9', 'UPI0000154552', 'A0A4D6FW43', 'A0A6M5E0N3', 'UPI0003F4A88A', 'UPI0015C551DC', 'UPI00090CD099', 'UPI0001A48A3B', 'K0NYR4', 'A0A1S5SHT2', 'UPI000B5ABD1C', 'UPI0001A48A9C', 'UPI000B5ABDA7', 'UPI000745F1F2', 'UPI00100D9C0B', 'UPI0001986816', 'UPI00019D0371', 'UPI00003E7814', 'UPI0001AF80B2', 'UPI000B5ABD86', 'UPI001262AC63', 'UPI0000458519', 'UPI00001556FA', 'UPI0019697F3C', 'UPI0014612D4E', 'UPI00137470B2', 'UPI001374707D', 'UPI000B5ABD65', 'UPI0003C712F7', 'B7UCZ6', 'A0A059PIU2', 'A0A5J6XWW3', 'UPI000C255C2F', 'UPI0001A48A37', 'A0A125NTU3', 'UPI0007414E7B', 'UPI001E1C222E', 'UPI00026BAC74', 'UPI000C81099F', 'UPI0000F58EE9', 'UPI0004D1BB55', 'UPI00001556F9', 'UPI000B5ABD6A', 'UPI0001AB2976', 'UPI001C9A2BD2', 'UPI00194ADA79', 'UPI000718AB89', 'A0A5J6CYR6', 'UPI0002A54BF8', 'UPI0001AB2977', 'UPI0001A48AA0', 'A0A7S7YQ80', 'UPI0002A54BB4', 'UPI000175260A', 'UPI000718AB72', 'UPI001CC3329B', 'UPI0001AF80B3', 'UPI001C717BE3', 'UPI0001AB2962', 'U6GSR1', 'B6F2F5', 'UPI0018FE02FF', 'UPI0007414E42', 'UPI00142F3CA5', 'UPI00026BAC78', 'UPI000B5ABD56', 'UPI000C255C22', 'UPI0013F595F6', 'UPI000153639E', 'UPI000259F50B', 'UPI00194ADAB5', 'UPI001D198A42', 'W8GG88', 'UPI0008595909', 'UPI0013F595FA', 'UPI00076F13AC', 'UPI00028BC7E8', 'UPI000B5ABD24', 'UPI00194ADAC1', 'UPI000B5ABD84', 'UPI0006AD099E', 'UPI000255FF86', 'UPI00194ADAD2', 'UPI0002250D5D', 'UPI000292D3B7', 'UPI00194ADA9D', 'UPI00194ADA50', 'UPI000C81095A', 'UPI00194ADA2C', 'UPI00194ADA88', 'UPI000277548C', 'UPI000255FF87', 'UPI00026BAC72', 'UPI00194ADA6B', 'A0A6I4B616', 'UPI000B5ABD91', 'UPI000C210336', 'UPI00194ADA7E', 'UPI00194ADA8C', 'UPI00194ADA9B', 'UPI0007A93650', 'UPI000255FFC7', 'UPI00194ADA74', 'UPI00004CCFAC', 'UPI001D0F4A03', 'UPI00194ADA57', 'UPI00004CCFAB', 'UPI0001AB296D', 'UPI00194ADAC5', 'UPI0001AB2961', 'UPI001C0A674F', 'UPI00194ADA7C', 'UPI0018D5A8E2', 'UPI00194ADAC0', 'UPI00004CCFAD', 'UPI00052E4983', 'UPI0010C76176', 'UPI0010C76139', 'UPI0003AFF93F', 'UPI00194ADA1D', 'UPI0001AB296E', 'UPI00194ADA90', 'UPI00025C06C1', 'A0A7S7YQ81', 'UPI0001F2FE6F', 'UPI00194ADAA6', 'UPI000225A884', 'I1YZZ9', 'UPI00194ADACE', 'UPI00051FF3EB', 'UPI00194ADA40', 'Q17105', 'UPI0001C67B2F', 'UPI000C810988', 'UPI000C6C8871', 'UPI000F044F5E', 'UPI0002AA2169', 'UPI00006833D0', 'UPI00027B2CE7', 'UPI000B5ABD7B', 'UPI001CEFB137', 'U6M5D0', 'UPI00194ADA6C', 'UPI000B5ABD22', 'UPI00194ADA5D', 'UPI000B5ABD9E', 'UPI0006526442', 'UPI00028BC767', 'UPI000E5587D0', 'UPI0009E0E32B', 'UPI00194ADA6A', 'UPI0001E30607', 'UPI00194ADAD4', 'UPI000B5ABDC0', 'UPI00078AF321', 'UPI0002B4DC8D', 'UPI00078AF337', 'UPI00082357E6', 'UPI00052E496B', 'UPI000B5ABD32', 'UPI001E1C2253', 'UPI0004F13DBB', 'UPI000E558788', 'UPI0010933757', 'UPI000E6AE39A', 'UPI00035AC992', 'UPI0006730543', 'Q17106', 'UPI0001A48AA1', 'UPI0001A48A3A', 'UPI000766E156', 'UPI00109337F7', 'UPI000B5ABD3F', 'UPI000E6AE36D', 'UPI00001CE540', 'UPI0002D257B2', 'UPI000745F1EF', 'UPI0008101732', 'UPI000789C189', 'UPI00187D6E10', 'UPI000B5ABDA4', 'Q6YGZ0', 'UPI0001AB295F', 'UPI0004DB2145', 'UPI00003E569F', 'UPI000B5ABD71', 'UPI00001CE53F', 'UPI0000D5BE2E', 'UPI0005841C9D', 'UPI00065B461A', 'UPI000B5ABD27', 'UPI0012B50503', 'UPI0010FFCD2E', 'UPI0010FFCD76', 'UPI00142F3CBD', 'UPI0002A54BFD', 'UPI000FF8754F', 'UPI000B5ABD2C', 'UPI00001BFB56', 'UPI00142602D2', 'UPI0001A48A9D', 'A0A348GST3', 'UPI000B5ABDA3', 'UPI0003B05A3D', 'UPI000FF8750E', 'UPI00022F83CA', 'UPI0005AECEC6', 'UPI000B5ABD99', 'UPI001CEFB14C', 'UPI000B5ABD96', 'UPI000B4966A5', 'UPI001A9A2C15', 'UPI000B4E83C3', 'UPI000B496691', 'UPI000B4E83E8', 'UPI0015882A93', 'UPI001CFD2F21', 'UPI000B4E83A9', 'UPI00085D8333', 'UPI000B496687', 'UPI000B4E8406', 'UPI0001DBAF3C', 'UPI000B49668F', 'UPI000B4E8377', 'UPI000B4966E4', 'UPI001D05C22D', 'UPI001D05C245', 'UPI00143F05B1', 'UPI000F45EFA4', 'UPI0009636567', 'UPI001D05C28B', 'UPI000745F231', 'D3TI87', 'UPI00178D0790', 'Q8WTC4', 'UPI0006AB76F2', 'UPI000745F1EA', 'Q8WTC6', 'UPI000745F236', 'Q8WTC5', 'UPI0001E3061D', 'UPI0001E3061E', 'UPI000745F255', 'UPI0005B2EBED', 'UPI0005C1213E', 'UPI0005B2EBAB', 'Q8WP95', 'UPI00142F3CE3', 'UPI000B4966D1', 'UPI0005CDCEA1', 'Q8WTC8', 'UPI0006418883', 'Q8WTD0', 'Q8WTC9', 'Q8WTC7', 'UPI00142F3CB0', 'UPI0005CDCEC1', 'UPI0005B2EBCD', 'UPI0019203BD9', 'UPI00022B80BB', 'UPI000745F249', 'UPI000FF874E1', 'UPI0005B2EBD1', 'UPI000D503263', 'UPI0006AB76EC', 'A0A5J6CYR9', 'A0A5J6CYN9', 'UPI001493DEA5', 'UPI00114D3DC4', 'UPI000C6F4430', 'A0A6N9ZS35', 'UPI0008FCAB2F', 'UPI000C6F4465', 'UPI001302D406', 'UPI000B5ABD1D', 'A0A1V4T0E4', 'A0A5J6CYI5', 'UPI000B5ABD49', 'Q6RYS7', 'A0A7M5UPI8', 'G1JSF2', 'A0A5J6CYK8', 'A0A7M5V224', 'A0A5J6CYV4', 'A0A7M5WQI9', 'A0A7M5XGN4', 'J9PGT0', 'A0A5J6CYT5', 'A0A7M5UKF9', 'G1JSF4', 'A0A7M5WWU1', 'G1JSF3', 'D7PM05', 'D7PM12', 'A0A7M5V407', 'D7PM10', 'D7PM06', 'D7PM04', 'A0A7M6DMT2', 'A0A7M5TUN8', 'A0A7M5WWD0', 'A0A7M5X361', 'J9PGG2', 'A0A5J6CYI7', 'UPI0000D5BE58', 'A0A5J6CYV7', 'UPI0000D5BE55', 'Q6RYS5', 'Q6RYS6', 'J9PIH6', 'A0A7M5TYI8', 'J9PJD5']\n",
      "34\n",
      "345\n",
      "34/34\n",
      "322/322\n"
     ]
    }
   ],
   "source": [
    "ids = aln['id']\n",
    "ids = [x for x in ids.apply(lambda x: x.split('/')[0]).values]\n",
    "\n",
    "print( [x.split('_')[1] for x in ids])\n",
    "# Sequences found in the very distant branch of the tree. \n",
    "outgroup = [\"UniRef100_Q6RYS7/1-232\",\n",
    "\"UniRef100_UPI000B5ABD49/2-234\",\n",
    "\"UniRef100_A0A5J6CYI5/1-229\",\n",
    "\"UniRef100_Q6RYS6/40-259\",\n",
    "\"UniRef100_Q6RYS5/2-223\",\n",
    "\"UniRef100_J9PIH6/6-230\",\n",
    "\"UniRef100_J9PJD5/7-231\",\n",
    "\"UniRef100_A0A7M5TYI8/7-231\",\n",
    "\"UniRef100_A0A5J6CYV7/1-229\",\n",
    "\"UniRef100_A0A5J6CYI7/1-229\",\n",
    "\"UniRef100_A0A5J6CYT5/1-229\",\n",
    "\"UniRef100_A0A5J6CYV4/1-229\",\n",
    "\"UniRef100_A0A5J6CYK8/1-229\",\n",
    "\"UniRef100_A0A7M5TUN8/3-229\",\n",
    "\"UniRef100_A0A7M5UKF9/3-229\",\n",
    "\"UniRef100_A0A7M5WQI9/3-229\",\n",
    "\"UniRef100_A0A7M5UPI8/3-229\",\n",
    "\"UniRef100_A0A7M5V224/3-229\",\n",
    "\"UniRef100_J9PGT0/3-229\",\n",
    "\"UniRef100_A0A7M5WWU1/3-229\",\n",
    "\"UniRef100_A0A7M5XGN4/3-229\",\n",
    "\"UniRef100_G1JSF2/3-231\",\n",
    "\"UniRef100_G1JSF3/3-231\",\n",
    "\"UniRef100_G1JSF4/3-231\",\n",
    "\"UniRef100_A0A7M5WWD0/7-234\",\n",
    "\"UniRef100_A0A7M5V407/7-234\",\n",
    "\"UniRef100_A0A7M6DMT2/54-281\",\n",
    "\"UniRef100_J9PGG2/7-234\",\n",
    "\"UniRef100_A0A7M5X361/50-277\",\n",
    "\"UniRef100_D7PM05/5-232\",\n",
    "\"UniRef100_D7PM10/5-232\",\n",
    "\"UniRef100_D7PM12/5-232\",\n",
    "\"UniRef100_D7PM04/5-232\",\n",
    "\"UniRef100_D7PM06/5-232\",]\n",
    "\n",
    "ID = 0 \n",
    "SPECIES = 1\n",
    "# in-group sequences \n",
    "outgroup = [x.split('/')[ID] for x in outgroup]\n",
    "print(len(outgroup))\n",
    "# FIX ME, querying for uniref seqs is still broken, skipping first uniprot entry \n",
    "ingroup = [x for x in ids[1:] if x not in outgroup]\n",
    "print(len(ingroup))\n",
    "\n",
    "uniprot_cols =  [\"id\", \"organism\"]\n",
    "outgroup_query = get_sequence_batch(outgroup, uniprot_cols, query_field = \"id\", database=\"uniref\")\n",
    "ingroup_query = get_sequence_batch(ingroup, uniprot_cols, query_field = \"id\", database=\"uniref\")\n",
    "\n",
    "outgroup_species = pd.Series([x[SPECIES] for x in outgroup_query]).value_counts()\n",
    "ingroup_species = pd.Series([x[SPECIES] for x in ingroup_query]).value_counts()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Removing synthetic sequences"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "360 380\n"
     ]
    }
   ],
   "source": [
    "synthetic_seqs = [x for x,y in ingroup_query if y == 'synthetic construct']\n",
    "\n",
    "no_syn_aln_seqs = []\n",
    "all_seqs = []\n",
    "for name, seq in zip(aln['id'], aln['sequence']):\n",
    "\n",
    "    if name.split('/')[0] not in synthetic_seqs:\n",
    "        x = Sequence(seq, name=name, gappy=True)\n",
    "        no_syn_aln_seqs.append(x)\n",
    "\n",
    "    x = Sequence(seq, name=name, gappy=True)\n",
    "    all_seqs.append(x)\n",
    "\n",
    "print(len(no_syn_aln_seqs), len(all_seqs))\n",
    "writeFastaFile(filename='no_synthetic_GFP_AEQVI_full_04-29-2022_b08.aln', seqs=no_syn_aln_seqs)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Difference at : 162\n"
     ]
    }
   ],
   "source": [
    "all_aln = Alignment(all_seqs)\n",
    "no_syn_aln = Alignment(no_syn_aln_seqs)\n",
    "\n",
    "con_all = all_aln.getConsensus() \n",
    "no_syn_con = no_syn_aln.getConsensus()\n",
    "count = 0 \n",
    "for x, y in zip(no_syn_con, con_all):\n",
    "    \n",
    "    if x != y:\n",
    "        print(f\"Difference at : {count}\")\n",
    "\n",
    "    count += 1\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Clytia hemisphaerica                      16\n",
       "Clytia gregaria (Gregarious jellyfish)     5\n",
       "Aequorea australis                         4\n",
       "Obelia sp. MH-2011                         3\n",
       "Phialidium sp. SL-2003                     2\n",
       "Aequorea victoria (Water jellyfish)        2\n",
       "Anthomedusae sp. SL-2003                   2\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "outgroup_species"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Aequorea victoria                                                                                                                                                                           189\n",
       "synthetic construct                                                                                                                                                                          20\n",
       "Homo sapiens                                                                                                                                                                                 14\n",
       "Rattus norvegicus                                                                                                                                                                            11\n",
       "Aequorea macrodactyla                                                                                                                                                                         9\n",
       "Aequorea victoria (Water jellyfish)                                                                                                                                                           7\n",
       "Escherichia coli                                                                                                                                                                              5\n",
       "Thermoanaerobacter sp. X513                                                                                                                                                                   3\n",
       "Entacmaea quadricolor                                                                                                                                                                         3\n",
       "Aequorea coerulescens                                                                                                                                                                         3\n",
       "Eimeria acervulina                                                                                                                                                                            2\n",
       "Mus musculus                                                                                                                                                                                  2\n",
       "Opsanus tau                                                                                                                                                                                   2\n",
       "cfp marker plasmid pWM1009                                                                                                                                                                    2\n",
       "Bacillus subtilis                                                                                                                                                                             1\n",
       "Nicotiana benthamiana; Azotobacter vinelandii                                                                                                                                                 1\n",
       "Eimeria maxima (Coccidian parasite)                                                                                                                                                           1\n",
       "Saccharomyces cerevisiae                                                                                                                                                                      1\n",
       "unclassified Pseudomonas                                                                                                                                                                      1\n",
       "Oryctolagus cuniculus                                                                                                                                                                         1\n",
       "Gallus gallus                                                                                                                                                                                 1\n",
       "Achromobacter denitrificans; Sinorhizobium fredii                                                                                                                                             1\n",
       "Helicobacter pylori                                                                                                                                                                           1\n",
       "Aspergillus niger                                                                                                                                                                             1\n",
       "Arabidopsis thaliana                                                                                                                                                                          1\n",
       "Hepacivirus hominis; Hepatitis C virus subtype 2a; Glutamicibacter soli; Pseudomonas fluorescens; Qipengyuania pelagi; Escherichia coli; Aequorea victoria                                    1\n",
       "Vibrio vulnificus                                                                                                                                                                             1\n",
       "Oceanospirillum multiglobuliferum                                                                                                                                                             1\n",
       "Camelus dromedarius                                                                                                                                                                           1\n",
       "Recombinant vesicular stomatitis Indiana virus rVSV-G/GFP; Vesicular stomatitis virus; Vibrio agarilyticus; Vesicular stomatitis Indiana virus strain Mudd-Summers; Shewanella intestini      1\n",
       "Babesia bovis                                                                                                                                                                                 1\n",
       "Aequorea victoria (Water jellyfish); Azotobacter vinelandii                                                                                                                                   1\n",
       "Azomonas agilis                                                                                                                                                                               1\n",
       "Human enterovirus 71 (EV71); Enterovirus A; Trichinella nativa                                                                                                                                1\n",
       "Eimeria acervulina (Coccidian parasite)                                                                                                                                                       1\n",
       "Microbacterium sp. AR7-10                                                                                                                                                                     1\n",
       "Aldersladia magnificus                                                                                                                                                                        1\n",
       "Hyphomicrobium sulfonivorans                                                                                                                                                                  1\n",
       "Achromobacter anxifer                                                                                                                                                                         1\n",
       "Rhizobium laguerreae; Bacteria                                                                                                                                                                1\n",
       "Human adenovirus C serotype 2 (HAdV-2)                                                                                                                                                        1\n",
       "Aequorea coerulescens (belt jellyfish)                                                                                                                                                        1\n",
       "Azotobacter vinelandii                                                                                                                                                                        1\n",
       "Bacilli                                                                                                                                                                                       1\n",
       "Escherichia coli O157:H7                                                                                                                                                                      1\n",
       "Sus scrofa; Bacillus anthracis                                                                                                                                                                1\n",
       "Blattella germanica                                                                                                                                                                           1\n",
       "Plasmodium falciparum 3D7                                                                                                                                                                     1\n",
       "unclassified Streptomyces                                                                                                                                                                     1\n",
       "Zaire ebolavirus; Sacbrood virus; synthetic recombinant Chikungunya virus strain LS3-GFP; Human immunodeficiency virus 1; Homo sapiens                                                        1\n",
       "Human respiratory syncytial virus                                                                                                                                                             1\n",
       "Streptomyces sp. FT05W; Streptomyces sp. ZEA17I; Ectothiorhodospiraceae bacterium WFHF3C12                                                                                                    1\n",
       "Aequorea australis                                                                                                                                                                            1\n",
       "Vicugna pacos                                                                                                                                                                                 1\n",
       "Pseudoalteromonas luteoviolacea                                                                                                                                                               1\n",
       "Bos taurus                                                                                                                                                                                    1\n",
       "Nicotiana benthamiana                                                                                                                                                                         1\n",
       "Recombinant vesicular stomatitis Indiana virus rVSV-G/GFP                                                                                                                                     1\n",
       "Homo sapiens; Bacillus subtilis; Mycobacterium tuberculosis                                                                                                                                   1\n",
       "Novosphingobium profundi                                                                                                                                                                      1\n",
       "Escherichia coli str. K-12 substr. MG1655; Pantoea agglomerans; Aeromonas australiensis                                                                                                       1\n",
       "Mesorhizobium japonicum                                                                                                                                                                       1\n",
       "Bacillus subtilis (strain 168)                                                                                                                                                                1\n",
       "Escherichia coli; Vibrio harveyi                                                                                                                                                              1\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ingroup_species"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Rooting tree"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Minimum ancestor deviation (MAD) will be used to root the tree as all sequences are assumed to be homologs. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Placing ancestors and extants in one file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Reading the alignment: /Users/sebs_mac/OneDrive - The University of Queensland/honours/data/gfp_alns/independent_runs/no_synthetic/alns/GFP_AEQVI_full_04-29-2022_b08_extants_no_syn.aln\n",
      "Checking for bad characters: ['B', 'J', 'X', 'Z']\n",
      "Performing one hot encoding\n",
      "Number of seqs: 360\n",
      "Reading the alignment: /Users/sebs_mac/OneDrive - The University of Queensland/honours/data/gfp_alns/independent_runs/no_synthetic/ancestors/auto_rooted//ancestors/run_1_ancestors.fa\n",
      "Checking for bad characters: ['B', 'J', 'X', 'Z']\n",
      "Number of seqs: 359\n",
      "Reading the alignment: /Users/sebs_mac/OneDrive - The University of Queensland/honours/data/gfp_alns/independent_runs/no_synthetic/ancestors/auto_rooted//ancestors/run_2_ancestors.fa\n",
      "Checking for bad characters: ['B', 'J', 'X', 'Z']\n",
      "Number of seqs: 359\n",
      "Reading the alignment: /Users/sebs_mac/OneDrive - The University of Queensland/honours/data/gfp_alns/independent_runs/no_synthetic/ancestors/auto_rooted//ancestors/run_3_ancestors.fa\n",
      "Checking for bad characters: ['B', 'J', 'X', 'Z']\n",
      "Number of seqs: 359\n",
      "Reading the alignment: /Users/sebs_mac/OneDrive - The University of Queensland/honours/data/gfp_alns/independent_runs/no_synthetic/ancestors/auto_rooted//ancestors/run_4_ancestors.fa\n",
      "Checking for bad characters: ['B', 'J', 'X', 'Z']\n",
      "Number of seqs: 359\n",
      "Reading the alignment: /Users/sebs_mac/OneDrive - The University of Queensland/honours/data/gfp_alns/independent_runs/no_synthetic/ancestors/auto_rooted//ancestors/run_5_ancestors.fa\n",
      "Checking for bad characters: ['B', 'J', 'X', 'Z']\n",
      "Number of seqs: 359\n",
      "Reading the alignment: /Users/sebs_mac/OneDrive - The University of Queensland/honours/data/gfp_alns/independent_runs/no_synthetic/ancestors/auto_rooted//ancestors/run_6_ancestors.fa\n",
      "Checking for bad characters: ['B', 'J', 'X', 'Z']\n",
      "Number of seqs: 359\n",
      "Reading the alignment: /Users/sebs_mac/OneDrive - The University of Queensland/honours/data/gfp_alns/independent_runs/no_synthetic/ancestors/auto_rooted//ancestors/run_7_ancestors.fa\n",
      "Checking for bad characters: ['B', 'J', 'X', 'Z']\n",
      "Number of seqs: 359\n",
      "Reading the alignment: /Users/sebs_mac/OneDrive - The University of Queensland/honours/data/gfp_alns/independent_runs/no_synthetic/ancestors/auto_rooted//ancestors/run_8_ancestors.fa\n",
      "Checking for bad characters: ['B', 'J', 'X', 'Z']\n",
      "Number of seqs: 359\n",
      "Reading the alignment: /Users/sebs_mac/OneDrive - The University of Queensland/honours/data/gfp_alns/independent_runs/no_synthetic/ancestors/auto_rooted//ancestors/run_9_ancestors.fa\n",
      "Checking for bad characters: ['B', 'J', 'X', 'Z']\n",
      "Number of seqs: 359\n",
      "Reading the alignment: /Users/sebs_mac/OneDrive - The University of Queensland/honours/data/gfp_alns/independent_runs/no_synthetic/ancestors/auto_rooted//ancestors/run_10_ancestors.fa\n",
      "Checking for bad characters: ['B', 'J', 'X', 'Z']\n",
      "Number of seqs: 359\n",
      "Reading the alignment: /Users/sebs_mac/OneDrive - The University of Queensland/honours/data/gfp_alns/independent_runs/no_synthetic/ancestors/auto_rooted//ancestors/run_12_ancestors.fa\n",
      "Checking for bad characters: ['B', 'J', 'X', 'Z']\n",
      "Number of seqs: 359\n",
      "Reading the alignment: /Users/sebs_mac/OneDrive - The University of Queensland/honours/data/gfp_alns/independent_runs/no_synthetic/ancestors/auto_rooted//ancestors/run_13_ancestors.fa\n",
      "Checking for bad characters: ['B', 'J', 'X', 'Z']\n",
      "Number of seqs: 359\n",
      "Reading the alignment: /Users/sebs_mac/OneDrive - The University of Queensland/honours/data/gfp_alns/independent_runs/no_synthetic/ancestors/auto_rooted//ancestors/run_14_ancestors.fa\n",
      "Checking for bad characters: ['B', 'J', 'X', 'Z']\n",
      "Number of seqs: 359\n",
      "Reading the alignment: /Users/sebs_mac/OneDrive - The University of Queensland/honours/data/gfp_alns/independent_runs/no_synthetic/ancestors/auto_rooted//ancestors/run_15_ancestors.fa\n",
      "Checking for bad characters: ['B', 'J', 'X', 'Z']\n",
      "Number of seqs: 359\n",
      "Reading the alignment: /Users/sebs_mac/OneDrive - The University of Queensland/honours/data/gfp_alns/independent_runs/no_synthetic/ancestors/auto_rooted//ancestors/run_16_ancestors.fa\n",
      "Checking for bad characters: ['B', 'J', 'X', 'Z']\n",
      "Number of seqs: 359\n",
      "Reading the alignment: /Users/sebs_mac/OneDrive - The University of Queensland/honours/data/gfp_alns/independent_runs/no_synthetic/ancestors/auto_rooted//ancestors/run_17_ancestors.fa\n",
      "Checking for bad characters: ['B', 'J', 'X', 'Z']\n",
      "Number of seqs: 359\n",
      "Reading the alignment: /Users/sebs_mac/OneDrive - The University of Queensland/honours/data/gfp_alns/independent_runs/no_synthetic/ancestors/auto_rooted//ancestors/run_18_ancestors.fa\n",
      "Checking for bad characters: ['B', 'J', 'X', 'Z']\n",
      "Number of seqs: 359\n",
      "Reading the alignment: /Users/sebs_mac/OneDrive - The University of Queensland/honours/data/gfp_alns/independent_runs/no_synthetic/ancestors/auto_rooted//ancestors/run_19_ancestors.fa\n",
      "Checking for bad characters: ['B', 'J', 'X', 'Z']\n",
      "Number of seqs: 359\n",
      "Reading the alignment: /Users/sebs_mac/OneDrive - The University of Queensland/honours/data/gfp_alns/independent_runs/no_synthetic/ancestors/auto_rooted//ancestors/run_20_ancestors.fa\n",
      "Checking for bad characters: ['B', 'J', 'X', 'Z']\n",
      "Number of seqs: 359\n",
      "Reading the alignment: /Users/sebs_mac/OneDrive - The University of Queensland/honours/data/gfp_alns/independent_runs/no_synthetic/ancestors/auto_rooted//ancestors/run_21_ancestors.fa\n",
      "Checking for bad characters: ['B', 'J', 'X', 'Z']\n",
      "Number of seqs: 359\n",
      "Reading the alignment: /Users/sebs_mac/OneDrive - The University of Queensland/honours/data/gfp_alns/independent_runs/no_synthetic/ancestors/auto_rooted//ancestors/run_22_ancestors.fa\n",
      "Checking for bad characters: ['B', 'J', 'X', 'Z']\n",
      "Number of seqs: 359\n",
      "Reading the alignment: /Users/sebs_mac/OneDrive - The University of Queensland/honours/data/gfp_alns/independent_runs/no_synthetic/ancestors/auto_rooted//ancestors/run_23_ancestors.fa\n",
      "Checking for bad characters: ['B', 'J', 'X', 'Z']\n",
      "Number of seqs: 359\n",
      "Reading the alignment: /Users/sebs_mac/OneDrive - The University of Queensland/honours/data/gfp_alns/independent_runs/no_synthetic/ancestors/auto_rooted//ancestors/run_24_ancestors.fa\n",
      "Checking for bad characters: ['B', 'J', 'X', 'Z']\n",
      "Number of seqs: 359\n",
      "Reading the alignment: /Users/sebs_mac/OneDrive - The University of Queensland/honours/data/gfp_alns/independent_runs/no_synthetic/ancestors/auto_rooted//ancestors/run_25_ancestors.fa\n",
      "Checking for bad characters: ['B', 'J', 'X', 'Z']\n",
      "Number of seqs: 359\n",
      "Reading the alignment: /Users/sebs_mac/OneDrive - The University of Queensland/honours/data/gfp_alns/independent_runs/no_synthetic/ancestors/auto_rooted//ancestors/run_26_ancestors.fa\n",
      "Checking for bad characters: ['B', 'J', 'X', 'Z']\n",
      "Number of seqs: 359\n",
      "Reading the alignment: /Users/sebs_mac/OneDrive - The University of Queensland/honours/data/gfp_alns/independent_runs/no_synthetic/ancestors/auto_rooted//ancestors/run_27_ancestors.fa\n",
      "Checking for bad characters: ['B', 'J', 'X', 'Z']\n",
      "Number of seqs: 359\n",
      "Reading the alignment: /Users/sebs_mac/OneDrive - The University of Queensland/honours/data/gfp_alns/independent_runs/no_synthetic/ancestors/auto_rooted//ancestors/run_28_ancestors.fa\n",
      "Checking for bad characters: ['B', 'J', 'X', 'Z']\n",
      "Number of seqs: 359\n",
      "Reading the alignment: /Users/sebs_mac/OneDrive - The University of Queensland/honours/data/gfp_alns/independent_runs/no_synthetic/ancestors/auto_rooted//ancestors/run_29_ancestors.fa\n",
      "Checking for bad characters: ['B', 'J', 'X', 'Z']\n",
      "Number of seqs: 359\n",
      "Reading the alignment: /Users/sebs_mac/OneDrive - The University of Queensland/honours/data/gfp_alns/independent_runs/no_synthetic/ancestors/auto_rooted//ancestors/run_30_ancestors.fa\n",
      "Checking for bad characters: ['B', 'J', 'X', 'Z']\n",
      "Number of seqs: 359\n",
      "Reading the alignment: /Users/sebs_mac/OneDrive - The University of Queensland/honours/data/gfp_alns/independent_runs/no_synthetic/ancestors/auto_rooted//ancestors/run_32_ancestors.fa\n",
      "Checking for bad characters: ['B', 'J', 'X', 'Z']\n",
      "Number of seqs: 359\n",
      "Reading the alignment: /Users/sebs_mac/OneDrive - The University of Queensland/honours/data/gfp_alns/independent_runs/no_synthetic/ancestors/auto_rooted//ancestors/run_33_ancestors.fa\n",
      "Checking for bad characters: ['B', 'J', 'X', 'Z']\n",
      "Number of seqs: 359\n",
      "Reading the alignment: /Users/sebs_mac/OneDrive - The University of Queensland/honours/data/gfp_alns/independent_runs/no_synthetic/ancestors/auto_rooted//ancestors/run_34_ancestors.fa\n",
      "Checking for bad characters: ['B', 'J', 'X', 'Z']\n",
      "Number of seqs: 359\n",
      "Reading the alignment: /Users/sebs_mac/OneDrive - The University of Queensland/honours/data/gfp_alns/independent_runs/no_synthetic/ancestors/auto_rooted//ancestors/run_35_ancestors.fa\n",
      "Checking for bad characters: ['B', 'J', 'X', 'Z']\n",
      "Number of seqs: 359\n",
      "Reading the alignment: /Users/sebs_mac/OneDrive - The University of Queensland/honours/data/gfp_alns/independent_runs/no_synthetic/ancestors/auto_rooted//ancestors/run_36_ancestors.fa\n",
      "Checking for bad characters: ['B', 'J', 'X', 'Z']\n",
      "Number of seqs: 359\n",
      "Reading the alignment: /Users/sebs_mac/OneDrive - The University of Queensland/honours/data/gfp_alns/independent_runs/no_synthetic/ancestors/auto_rooted//ancestors/run_37_ancestors.fa\n",
      "Checking for bad characters: ['B', 'J', 'X', 'Z']\n",
      "Number of seqs: 359\n",
      "Reading the alignment: /Users/sebs_mac/OneDrive - The University of Queensland/honours/data/gfp_alns/independent_runs/no_synthetic/ancestors/auto_rooted//ancestors/run_38_ancestors.fa\n",
      "Checking for bad characters: ['B', 'J', 'X', 'Z']\n",
      "Number of seqs: 359\n",
      "Reading the alignment: /Users/sebs_mac/OneDrive - The University of Queensland/honours/data/gfp_alns/independent_runs/no_synthetic/ancestors/auto_rooted//ancestors/run_39_ancestors.fa\n",
      "Checking for bad characters: ['B', 'J', 'X', 'Z']\n",
      "Number of seqs: 359\n",
      "Reading the alignment: /Users/sebs_mac/OneDrive - The University of Queensland/honours/data/gfp_alns/independent_runs/no_synthetic/ancestors/auto_rooted//ancestors/run_40_ancestors.fa\n",
      "Checking for bad characters: ['B', 'J', 'X', 'Z']\n",
      "Number of seqs: 359\n",
      "Reading the alignment: /Users/sebs_mac/OneDrive - The University of Queensland/honours/data/gfp_alns/independent_runs/no_synthetic/ancestors/auto_rooted//ancestors/run_41_ancestors.fa\n",
      "Checking for bad characters: ['B', 'J', 'X', 'Z']\n",
      "Number of seqs: 359\n",
      "Reading the alignment: /Users/sebs_mac/OneDrive - The University of Queensland/honours/data/gfp_alns/independent_runs/no_synthetic/ancestors/auto_rooted//ancestors/run_42_ancestors.fa\n",
      "Checking for bad characters: ['B', 'J', 'X', 'Z']\n",
      "Number of seqs: 359\n",
      "Reading the alignment: /Users/sebs_mac/OneDrive - The University of Queensland/honours/data/gfp_alns/independent_runs/no_synthetic/ancestors/auto_rooted//ancestors/run_43_ancestors.fa\n",
      "Checking for bad characters: ['B', 'J', 'X', 'Z']\n",
      "Number of seqs: 359\n",
      "Reading the alignment: /Users/sebs_mac/OneDrive - The University of Queensland/honours/data/gfp_alns/independent_runs/no_synthetic/ancestors/auto_rooted//ancestors/run_44_ancestors.fa\n",
      "Checking for bad characters: ['B', 'J', 'X', 'Z']\n",
      "Number of seqs: 359\n",
      "Reading the alignment: /Users/sebs_mac/OneDrive - The University of Queensland/honours/data/gfp_alns/independent_runs/no_synthetic/ancestors/auto_rooted//ancestors/run_45_ancestors.fa\n",
      "Checking for bad characters: ['B', 'J', 'X', 'Z']\n",
      "Number of seqs: 359\n",
      "Reading the alignment: /Users/sebs_mac/OneDrive - The University of Queensland/honours/data/gfp_alns/independent_runs/no_synthetic/ancestors/auto_rooted//ancestors/run_46_ancestors.fa\n",
      "Checking for bad characters: ['B', 'J', 'X', 'Z']\n",
      "Number of seqs: 359\n",
      "Reading the alignment: /Users/sebs_mac/OneDrive - The University of Queensland/honours/data/gfp_alns/independent_runs/no_synthetic/ancestors/auto_rooted//ancestors/run_47_ancestors.fa\n",
      "Checking for bad characters: ['B', 'J', 'X', 'Z']\n",
      "Number of seqs: 359\n",
      "Reading the alignment: /Users/sebs_mac/OneDrive - The University of Queensland/honours/data/gfp_alns/independent_runs/no_synthetic/ancestors/auto_rooted//ancestors/run_49_ancestors.fa\n",
      "Checking for bad characters: ['B', 'J', 'X', 'Z']\n",
      "Number of seqs: 359\n",
      "Reading the alignment: /Users/sebs_mac/OneDrive - The University of Queensland/honours/data/gfp_alns/independent_runs/no_synthetic/ancestors/auto_rooted//ancestors/run_50_ancestors.fa\n",
      "Checking for bad characters: ['B', 'J', 'X', 'Z']\n",
      "Number of seqs: 359\n",
      "Reading the alignment: /Users/sebs_mac/OneDrive - The University of Queensland/honours/data/gfp_alns/independent_runs/no_synthetic/ancestors/auto_rooted//ancestors/run_51_ancestors.fa\n",
      "Checking for bad characters: ['B', 'J', 'X', 'Z']\n",
      "Number of seqs: 359\n",
      "Reading the alignment: /Users/sebs_mac/OneDrive - The University of Queensland/honours/data/gfp_alns/independent_runs/no_synthetic/ancestors/auto_rooted//ancestors/run_52_ancestors.fa\n",
      "Checking for bad characters: ['B', 'J', 'X', 'Z']\n",
      "Number of seqs: 359\n",
      "Reading the alignment: /Users/sebs_mac/OneDrive - The University of Queensland/honours/data/gfp_alns/independent_runs/no_synthetic/ancestors/auto_rooted//ancestors/run_53_ancestors.fa\n",
      "Checking for bad characters: ['B', 'J', 'X', 'Z']\n",
      "Number of seqs: 359\n",
      "Reading the alignment: /Users/sebs_mac/OneDrive - The University of Queensland/honours/data/gfp_alns/independent_runs/no_synthetic/ancestors/auto_rooted//ancestors/run_54_ancestors.fa\n",
      "Checking for bad characters: ['B', 'J', 'X', 'Z']\n",
      "Number of seqs: 359\n",
      "Reading the alignment: /Users/sebs_mac/OneDrive - The University of Queensland/honours/data/gfp_alns/independent_runs/no_synthetic/ancestors/auto_rooted//ancestors/run_55_ancestors.fa\n",
      "Checking for bad characters: ['B', 'J', 'X', 'Z']\n",
      "Number of seqs: 359\n",
      "Reading the alignment: /Users/sebs_mac/OneDrive - The University of Queensland/honours/data/gfp_alns/independent_runs/no_synthetic/ancestors/auto_rooted//ancestors/run_56_ancestors.fa\n",
      "Checking for bad characters: ['B', 'J', 'X', 'Z']\n",
      "Number of seqs: 359\n",
      "Reading the alignment: /Users/sebs_mac/OneDrive - The University of Queensland/honours/data/gfp_alns/independent_runs/no_synthetic/ancestors/auto_rooted//ancestors/run_57_ancestors.fa\n",
      "Checking for bad characters: ['B', 'J', 'X', 'Z']\n",
      "Number of seqs: 359\n",
      "Reading the alignment: /Users/sebs_mac/OneDrive - The University of Queensland/honours/data/gfp_alns/independent_runs/no_synthetic/ancestors/auto_rooted//ancestors/run_58_ancestors.fa\n",
      "Checking for bad characters: ['B', 'J', 'X', 'Z']\n",
      "Number of seqs: 359\n",
      "Reading the alignment: /Users/sebs_mac/OneDrive - The University of Queensland/honours/data/gfp_alns/independent_runs/no_synthetic/ancestors/auto_rooted//ancestors/run_59_ancestors.fa\n",
      "Checking for bad characters: ['B', 'J', 'X', 'Z']\n",
      "Number of seqs: 359\n",
      "Reading the alignment: /Users/sebs_mac/OneDrive - The University of Queensland/honours/data/gfp_alns/independent_runs/no_synthetic/ancestors/auto_rooted//ancestors/run_61_ancestors.fa\n",
      "Checking for bad characters: ['B', 'J', 'X', 'Z']\n",
      "Number of seqs: 359\n",
      "Reading the alignment: /Users/sebs_mac/OneDrive - The University of Queensland/honours/data/gfp_alns/independent_runs/no_synthetic/ancestors/auto_rooted//ancestors/run_62_ancestors.fa\n",
      "Checking for bad characters: ['B', 'J', 'X', 'Z']\n",
      "Number of seqs: 359\n",
      "Reading the alignment: /Users/sebs_mac/OneDrive - The University of Queensland/honours/data/gfp_alns/independent_runs/no_synthetic/ancestors/auto_rooted//ancestors/run_64_ancestors.fa\n",
      "Checking for bad characters: ['B', 'J', 'X', 'Z']\n",
      "Number of seqs: 359\n",
      "Reading the alignment: /Users/sebs_mac/OneDrive - The University of Queensland/honours/data/gfp_alns/independent_runs/no_synthetic/ancestors/auto_rooted//ancestors/run_65_ancestors.fa\n",
      "Checking for bad characters: ['B', 'J', 'X', 'Z']\n",
      "Number of seqs: 359\n",
      "Reading the alignment: /Users/sebs_mac/OneDrive - The University of Queensland/honours/data/gfp_alns/independent_runs/no_synthetic/ancestors/auto_rooted//ancestors/run_66_ancestors.fa\n",
      "Checking for bad characters: ['B', 'J', 'X', 'Z']\n",
      "Number of seqs: 359\n",
      "Reading the alignment: /Users/sebs_mac/OneDrive - The University of Queensland/honours/data/gfp_alns/independent_runs/no_synthetic/ancestors/auto_rooted//ancestors/run_67_ancestors.fa\n",
      "Checking for bad characters: ['B', 'J', 'X', 'Z']\n",
      "Number of seqs: 359\n",
      "Reading the alignment: /Users/sebs_mac/OneDrive - The University of Queensland/honours/data/gfp_alns/independent_runs/no_synthetic/ancestors/auto_rooted//ancestors/run_68_ancestors.fa\n",
      "Checking for bad characters: ['B', 'J', 'X', 'Z']\n",
      "Number of seqs: 359\n",
      "Reading the alignment: /Users/sebs_mac/OneDrive - The University of Queensland/honours/data/gfp_alns/independent_runs/no_synthetic/ancestors/auto_rooted//ancestors/run_69_ancestors.fa\n",
      "Checking for bad characters: ['B', 'J', 'X', 'Z']\n",
      "Number of seqs: 359\n",
      "Reading the alignment: /Users/sebs_mac/OneDrive - The University of Queensland/honours/data/gfp_alns/independent_runs/no_synthetic/ancestors/auto_rooted//ancestors/run_70_ancestors.fa\n",
      "Checking for bad characters: ['B', 'J', 'X', 'Z']\n",
      "Number of seqs: 359\n",
      "Reading the alignment: /Users/sebs_mac/OneDrive - The University of Queensland/honours/data/gfp_alns/independent_runs/no_synthetic/ancestors/auto_rooted//ancestors/run_71_ancestors.fa\n",
      "Checking for bad characters: ['B', 'J', 'X', 'Z']\n",
      "Number of seqs: 359\n",
      "Reading the alignment: /Users/sebs_mac/OneDrive - The University of Queensland/honours/data/gfp_alns/independent_runs/no_synthetic/ancestors/auto_rooted//ancestors/run_72_ancestors.fa\n",
      "Checking for bad characters: ['B', 'J', 'X', 'Z']\n",
      "Number of seqs: 359\n",
      "Reading the alignment: /Users/sebs_mac/OneDrive - The University of Queensland/honours/data/gfp_alns/independent_runs/no_synthetic/ancestors/auto_rooted//ancestors/run_73_ancestors.fa\n",
      "Checking for bad characters: ['B', 'J', 'X', 'Z']\n",
      "Number of seqs: 359\n",
      "Reading the alignment: /Users/sebs_mac/OneDrive - The University of Queensland/honours/data/gfp_alns/independent_runs/no_synthetic/ancestors/auto_rooted//ancestors/run_74_ancestors.fa\n",
      "Checking for bad characters: ['B', 'J', 'X', 'Z']\n",
      "Number of seqs: 359\n",
      "Reading the alignment: /Users/sebs_mac/OneDrive - The University of Queensland/honours/data/gfp_alns/independent_runs/no_synthetic/ancestors/auto_rooted//ancestors/run_75_ancestors.fa\n",
      "Checking for bad characters: ['B', 'J', 'X', 'Z']\n",
      "Number of seqs: 359\n",
      "Reading the alignment: /Users/sebs_mac/OneDrive - The University of Queensland/honours/data/gfp_alns/independent_runs/no_synthetic/ancestors/auto_rooted//ancestors/run_76_ancestors.fa\n",
      "Checking for bad characters: ['B', 'J', 'X', 'Z']\n",
      "Number of seqs: 359\n",
      "Reading the alignment: /Users/sebs_mac/OneDrive - The University of Queensland/honours/data/gfp_alns/independent_runs/no_synthetic/ancestors/auto_rooted//ancestors/run_77_ancestors.fa\n",
      "Checking for bad characters: ['B', 'J', 'X', 'Z']\n",
      "Number of seqs: 359\n",
      "Reading the alignment: /Users/sebs_mac/OneDrive - The University of Queensland/honours/data/gfp_alns/independent_runs/no_synthetic/ancestors/auto_rooted//ancestors/run_79_ancestors.fa\n",
      "Checking for bad characters: ['B', 'J', 'X', 'Z']\n",
      "Number of seqs: 359\n",
      "Reading the alignment: /Users/sebs_mac/OneDrive - The University of Queensland/honours/data/gfp_alns/independent_runs/no_synthetic/ancestors/auto_rooted//ancestors/run_80_ancestors.fa\n",
      "Checking for bad characters: ['B', 'J', 'X', 'Z']\n",
      "Number of seqs: 359\n",
      "Reading the alignment: /Users/sebs_mac/OneDrive - The University of Queensland/honours/data/gfp_alns/independent_runs/no_synthetic/ancestors/auto_rooted//ancestors/run_81_ancestors.fa\n",
      "Checking for bad characters: ['B', 'J', 'X', 'Z']\n",
      "Number of seqs: 359\n",
      "Reading the alignment: /Users/sebs_mac/OneDrive - The University of Queensland/honours/data/gfp_alns/independent_runs/no_synthetic/ancestors/auto_rooted//ancestors/run_82_ancestors.fa\n",
      "Checking for bad characters: ['B', 'J', 'X', 'Z']\n",
      "Number of seqs: 359\n",
      "Reading the alignment: /Users/sebs_mac/OneDrive - The University of Queensland/honours/data/gfp_alns/independent_runs/no_synthetic/ancestors/auto_rooted//ancestors/run_83_ancestors.fa\n",
      "Checking for bad characters: ['B', 'J', 'X', 'Z']\n",
      "Number of seqs: 359\n",
      "Reading the alignment: /Users/sebs_mac/OneDrive - The University of Queensland/honours/data/gfp_alns/independent_runs/no_synthetic/ancestors/auto_rooted//ancestors/run_84_ancestors.fa\n",
      "Checking for bad characters: ['B', 'J', 'X', 'Z']\n",
      "Number of seqs: 359\n",
      "Reading the alignment: /Users/sebs_mac/OneDrive - The University of Queensland/honours/data/gfp_alns/independent_runs/no_synthetic/ancestors/auto_rooted//ancestors/run_85_ancestors.fa\n",
      "Checking for bad characters: ['B', 'J', 'X', 'Z']\n",
      "Number of seqs: 359\n",
      "Reading the alignment: /Users/sebs_mac/OneDrive - The University of Queensland/honours/data/gfp_alns/independent_runs/no_synthetic/ancestors/auto_rooted//ancestors/run_86_ancestors.fa\n",
      "Checking for bad characters: ['B', 'J', 'X', 'Z']\n",
      "Number of seqs: 359\n",
      "Reading the alignment: /Users/sebs_mac/OneDrive - The University of Queensland/honours/data/gfp_alns/independent_runs/no_synthetic/ancestors/auto_rooted//ancestors/run_87_ancestors.fa\n",
      "Checking for bad characters: ['B', 'J', 'X', 'Z']\n",
      "Number of seqs: 359\n",
      "Reading the alignment: /Users/sebs_mac/OneDrive - The University of Queensland/honours/data/gfp_alns/independent_runs/no_synthetic/ancestors/auto_rooted//ancestors/run_88_ancestors.fa\n",
      "Checking for bad characters: ['B', 'J', 'X', 'Z']\n",
      "Number of seqs: 359\n",
      "Reading the alignment: /Users/sebs_mac/OneDrive - The University of Queensland/honours/data/gfp_alns/independent_runs/no_synthetic/ancestors/auto_rooted//ancestors/run_89_ancestors.fa\n",
      "Checking for bad characters: ['B', 'J', 'X', 'Z']\n",
      "Number of seqs: 359\n",
      "Reading the alignment: /Users/sebs_mac/OneDrive - The University of Queensland/honours/data/gfp_alns/independent_runs/no_synthetic/ancestors/auto_rooted//ancestors/run_90_ancestors.fa\n",
      "Checking for bad characters: ['B', 'J', 'X', 'Z']\n",
      "Number of seqs: 359\n",
      "Reading the alignment: /Users/sebs_mac/OneDrive - The University of Queensland/honours/data/gfp_alns/independent_runs/no_synthetic/ancestors/auto_rooted//ancestors/run_91_ancestors.fa\n",
      "Checking for bad characters: ['B', 'J', 'X', 'Z']\n",
      "Number of seqs: 359\n",
      "Reading the alignment: /Users/sebs_mac/OneDrive - The University of Queensland/honours/data/gfp_alns/independent_runs/no_synthetic/ancestors/auto_rooted//ancestors/run_92_ancestors.fa\n",
      "Checking for bad characters: ['B', 'J', 'X', 'Z']\n",
      "Number of seqs: 359\n",
      "Reading the alignment: /Users/sebs_mac/OneDrive - The University of Queensland/honours/data/gfp_alns/independent_runs/no_synthetic/ancestors/auto_rooted//ancestors/run_93_ancestors.fa\n",
      "Checking for bad characters: ['B', 'J', 'X', 'Z']\n",
      "Number of seqs: 359\n",
      "Reading the alignment: /Users/sebs_mac/OneDrive - The University of Queensland/honours/data/gfp_alns/independent_runs/no_synthetic/ancestors/auto_rooted//ancestors/run_94_ancestors.fa\n",
      "Checking for bad characters: ['B', 'J', 'X', 'Z']\n",
      "Number of seqs: 359\n",
      "Reading the alignment: /Users/sebs_mac/OneDrive - The University of Queensland/honours/data/gfp_alns/independent_runs/no_synthetic/ancestors/auto_rooted//ancestors/run_95_ancestors.fa\n",
      "Checking for bad characters: ['B', 'J', 'X', 'Z']\n",
      "Number of seqs: 359\n",
      "Reading the alignment: /Users/sebs_mac/OneDrive - The University of Queensland/honours/data/gfp_alns/independent_runs/no_synthetic/ancestors/auto_rooted//ancestors/run_96_ancestors.fa\n",
      "Checking for bad characters: ['B', 'J', 'X', 'Z']\n",
      "Number of seqs: 359\n",
      "Reading the alignment: /Users/sebs_mac/OneDrive - The University of Queensland/honours/data/gfp_alns/independent_runs/no_synthetic/ancestors/auto_rooted//ancestors/run_97_ancestors.fa\n",
      "Checking for bad characters: ['B', 'J', 'X', 'Z']\n",
      "Number of seqs: 359\n",
      "Reading the alignment: /Users/sebs_mac/OneDrive - The University of Queensland/honours/data/gfp_alns/independent_runs/no_synthetic/ancestors/auto_rooted//ancestors/run_98_ancestors.fa\n",
      "Checking for bad characters: ['B', 'J', 'X', 'Z']\n",
      "Number of seqs: 359\n",
      "Reading the alignment: /Users/sebs_mac/OneDrive - The University of Queensland/honours/data/gfp_alns/independent_runs/no_synthetic/ancestors/auto_rooted//ancestors/run_99_ancestors.fa\n",
      "Checking for bad characters: ['B', 'J', 'X', 'Z']\n",
      "Number of seqs: 359\n",
      "Reading the alignment: /Users/sebs_mac/OneDrive - The University of Queensland/honours/data/gfp_alns/independent_runs/no_synthetic/ancestors/auto_rooted//ancestors/run_100_ancestors.fa\n",
      "Checking for bad characters: ['B', 'J', 'X', 'Z']\n",
      "Number of seqs: 359\n"
     ]
    }
   ],
   "source": [
    "ANC_DIR = '/Users/sebs_mac/OneDrive - The University of Queensland/honours/data/gfp_alns/independent_runs/no_synthetic/ancestors/auto_rooted/'\n",
    "DATA_DIR = '/Users/sebs_mac/OneDrive - The University of Queensland/honours/data/gfp_alns/independent_runs/no_synthetic/alns/'\n",
    "\n",
    "# trees rejected by AU test or failed run\n",
    "bad_trees = [11, 31, 48, 60, 78, 63]\n",
    "# don't use ancestors from the outgroup. \n",
    "#bad_ancestors = ['N' + str(i) for i in range(0,34)]\n",
    "anc_extants = []\n",
    "anc_only = []\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "# EXTANTS \n",
    "aln = st.read_aln_file(DATA_DIR + 'GFP_AEQVI_full_04-29-2022_b08_extants_no_syn.aln')\n",
    "for name, seq in zip(aln['id'], aln['sequence']):\n",
    "    x = Sequence(seq, name=name, gappy=True)\n",
    "    anc_extants.append(x)\n",
    "\n",
    "# ANCESTORS\n",
    "for i in range(1,101):\n",
    "\n",
    "    if i in bad_trees:\n",
    "        continue \n",
    "\n",
    "    aln = st.read_aln_file(ANC_DIR + f'/ancestors/run_{i}_ancestors.fa', encode=False)\n",
    "    for name, seq in zip(aln['id'], aln['sequence']):\n",
    "      \n",
    "        seq_name = name + f\"_tree_{i}\"\n",
    "        x = Sequence(seq, name=seq_name, gappy=True)\n",
    "        anc_extants.append(x)\n",
    "        anc_only.append(x)\n",
    "\n",
    "OUT_DIR = \"/Users/sebs_mac/OneDrive - The University of Queensland/honours/data/gfp_alns/independent_runs/no_synthetic/alns/\"\n",
    "writeFastaFile(OUT_DIR + 'GFP_AEQVI_full_04-29-2022_b08_ancestors_extants_no_syn.aln', anc_extants)\n",
    "writeFastaFile(OUT_DIR + 'GFP_AEQVI_full_04-29-2022_b08_ancestors_no_syn.aln', anc_only)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "embed",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.18"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
